src/repository/mod.rs

```diff
@@ -3,3 +3,57 @@ mod entities;
 pub mod interface;
 pub mod mongo;
 pub mod sql;
+use crate::settings::Setup;
+use std::{io::Result, time::Duration};
+
+#[derive(Clone, Debug)]
+pub enum Repository {
+    Mongo(mongo::WarehouseMongo),
+    Sql(sql::WarehouseSql),
+}
+
+impl Repository {
+    pub async fn new(s: &Setup) -> Result<Self> {
+        let conn_str = s.get_connection_str();
+
+        if conn_str.contains("mongodb") {
+            let m = mongo::WarehouseMongo::new(&conn_str).await?;
+            return Ok(Self::Mongo(m));
+        }
+        if !conn_str.is_empty() {
+            let s = sql::WarehouseSql::new(sql::DatabaseStorage::Path(conn_str)).await?;
+            return Ok(Self::Sql(s));
+        }
+        let s = sql::WarehouseSql::new(sql::DatabaseStorage::Ram).await?;
+        Ok(Self::Sql(s))
+    }
+}
+
+impl interface::RepositoryProvider for Repository {
+    async fn migrate(&self) -> Result<()> {
+        match &self {
+            Repository::Mongo(r) => r.migrate().await,
+            Repository::Sql(r) => r.migrate().await,
+        }
```

Comment:
Question: If we await everything, what benefit we have of async?

Reply:
First the `actix-web` works based on actors that are threaded. When threads are exhausted (you can run effectively in parallel as many threads as CPU allows, or even less in practice), then all you can do is switch context (do async execution). In async when one thing is waiting then the other can execute. So the benefit of awaiting things that can wait for the system, DB, and external service is that while one is waiting another can perform. The next benefit is that async / await gives you a very well-designed abstraction over waiting for a response and timing out at some point without blocking.

Reply:
Imagine that you are driving a car to work. Let's assume there is no awaiting in your actions.  When you stop at a red light you are waiting, you want to go to your radio and pick up the radio station, but you cannot switch the context from driving because you are in no await mode, so you will set the radio station only after you arrive to the work. The more things that may block execution we await, the better we use CPU time.